File: src/main/java/net/sf/prayer/Context.java
package net.sf.prayer;

import java.util.*;

public class Context {

    private OpStack stack;
    private Map<String, Object> globals;
    private ArrayList<Object[]> locals;
    private Object[] curLocals;
    private int callDepth;
    
    public Context() {
        stack = new OpStack();
        globals = new HashMap<>();
        locals = new ArrayList<>();
        callDepth = 0;
    }
    
    public void increaseCallDepth() {
        while (locals.size() <= callDepth) {
            locals.add(new Object[26]);
        }
        curLocals = locals.get(callDepth);
        callDepth++;
    }
    
    public void decreaseCallDepth() {
        callDepth--;
        curLocals = (callDepth > 0) ? locals.get(callDepth - 1) : null;
    }
    
    public OpStack getStack() {
        return stack;
    }
    
    public void assignVar(String name, Object v) {
        if (name.length() > 1) {
            globals.put(name, v);
        } else {
            curLocals[name.charAt(0) - 'a'] = v;
        }
    }
    
    public Object fetchVar(String name) {
        Object res;
        if (name.length() > 1) {
            res = globals.get(name);
        } else {
            res = curLocals[name.charAt(0) - 'a'];
        }
        return res;
    }
    
}

File: javascript/context.js
function Context() {
    this.stack = [];
    this.globals = [];
    this.locals = [];
    this.curLocals = null;
    this.callDepth = 0;
}

Context.prototype.increaseCallDepth = function() {
    while (this.locals.length <= this.callDepth) {
        this.locals.push([]);
    }
    this.curLocals = this.locals[this.callDepth];
    this.callDepth++;
}

Context.prototype.decreaseCallDepth = function() {
    this.callDepth--;
    this.curLocals = (this.callDepth > 0) ? this.locals[this.callDepth - 1] : null;
}

Context.prototype.assignVar = function(name, v) {
    if (name.length > 1) {
        this.globals[name] = v;
    } else {
        this.curLocals[name] = v;
    }
}

Context.prototype.fetchVar = function(name) {
    if (name.length > 1) {
        return this.globals[name];
    } else {
        return this.curLocals[name];
    }
}
